### PR TITLE
perf(expr): complete expression benchmark framework

### LINKED ISSUE #3524: Introduce expression benchmark framework
🤔

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

This PR completes the micro-benchmark framework for expressions.

It can be run with:
```sh
cd src/expr
# run all benches
cargo bench --bench expr -- --quick
# list all benches
cargo bench --bench expr -- --list
# run specified benches
cargo bench --bench expr -- --quick "add\(int32,int32\)"
```

The detailed bench results have been updated to #6868.
Here is a statistical overview of all expressions:

截屏2022-12-20 21 13 24

To enumerate all valid expressions, we utilize the function signature maps defined in the frontend. We moved them into the expr crate in order to avoid dependency on the frontend thus reducing the compilation time.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
resolve #3524

### CODE REVIEW

File: src/expr/benches/expr.rs [hunk 1]
```diff
@@ -20,63 +20,301 @@
...
+                .collect_vec(),
+        );
+        let expr = match build_from_prost(&prost) {
+            Ok(expr) => expr,
+            Err(e) => {
+                println!("{e}");
+                continue;
+            }
+        };
+        c.bench_function(&sig.to_string_no_return(), |bencher| {
```
Comment (BugenZhao) [inline src/expr/benches/expr.rs:248]: For the first time, I know this can be dynamic. 😄

**Review by BugenZhao — COMMENTED** [review]
The `array` and `benches` LGTM! 🚀

**Comment by TennyZhuang** [comment]
It'll be better if we can run the benchmark of every function separately, e.g., like [how TiDB does](https://github.com/pingcap/tidb/blob/130568716ef49b71746d1d14ffe2b2a216961c35/Makefile#L229).

The most use case is that we may optimize one specified function locally, but we don't want to run the full benchmark. It's too heavy.

**Comment by TennyZhuang** [comment]
We can only cover a subset of all functions and extend them later on demand.

**Comment by wangrunji0408** [comment]
> It'll be better if we can run the benchmark of every function separately, e.g., like [how TiDB does](https://github.com/pingcap/tidb/blob/130568716ef49b71746d1d14ffe2b2a216961c35/Makefile#L229).
> 
> The most use case is that we may optimize one specified function locally, but we don't want to run the full benchmark. It's too heavy.

Of course we can run them separately: 😀

```
cargo bench -- --quick [filter]
cargo bench -- --quick is_false
cargo bench -- --quick "add\(int32,int32\)"
```

**Comment by TennyZhuang** [comment]
> > It'll be better if we can run the benchmark of every function separately, e.g., like [how TiDB does](https://github.com/pingcap/tidb/blob/130568716ef49b71746d1d14ffe2b2a216961c35/Makefile#L229).
> > The most use case is that we may optimize one specified function locally, but we don't want to run the full benchmark. It's too heavy.
> 
> Of course we can run them separately: 😀
> 
> ```
> cargo bench -- --quick [filter]
> cargo bench -- --quick is_false
> cargo bench -- --quick "add\(int32,int32\)"
> ```

Great work!

**Review by lmatz — APPROVED** [review]

**Pull request merged**